src/MiniProfiler.AspNetCore.Mvc/MvcDiagnosticListener.cs

```diff
@@ -0,0 +1,166 @@
+﻿#if NETCOREAPP3_0
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.Diagnostics;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.Razor;
+using StackExchange.Profiling.Internal;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+
+namespace StackExchange.Profiling.Data
+{
+    /// <summary>
+    /// Diagnostic listener for Microsoft.AspNetCore.Mvc.* events
+    /// </summary>
+    /// <remarks>
+    /// See:
+    /// - https://github.com/dotnet/aspnetcore/blob/v3.0.0/src/Mvc/Mvc.Core/src/MvcCoreDiagnosticListenerExtensions.cs
+    /// - https://github.com/dotnet/aspnetcore/blob/v3.0.0/src/Mvc/Mvc.Core/src/Diagnostics/MvcDiagnostics.cs
+    /// - https://github.com/dotnet/aspnetcore/blob/v3.0.0/src/Mvc/Mvc.Razor/src/Diagnostics/MvcDiagnostics.cs
+    /// </remarks>
+    public class MvcDiagnosticListener : IMiniProfilerDiagnosticListener
+    {
+        /// <summary>
+        /// Diagnostic Listener name to handle.
+        /// </summary>
+        public string ListenerName => "Microsoft.AspNetCore";
+
+        /// <summary>
+        /// Notifies the observer that the provider has finished sending push-based notifications.
+        /// </summary>
+        public void OnCompleted() { }
+
+        /// <summary>
+        /// Notifies the observer that the provider has experienced an error condition.
+        /// </summary>
+        /// <param name="error">An object that provides additional information about the error.</param>
+        public void OnError(Exception error) => Trace.WriteLine(error);
+
+        private static string GetName(string label, ActionDescriptor descriptor)
+        {
+            var controller = descriptor.RouteValues.TryGetValue("controller", out var c) ? c : "UnknownController";
+            var action = descriptor.RouteValues.TryGetValue("action", out var a) ? a : "UnknownAction";
+
+            // TODO: Don't allocate this string more than once
+            return label + ": " + controller + "/" + action;
+        }
+
+        private static string GetName(IActionResult result) => result switch
+        {
+            ViewResult vr => vr.ViewName.HasValue() ? "View: " + vr.ViewName : "ViewResult",
+            ContentResult cr => cr.ContentType.HasValue() ? "Content: " + cr.ContentType : "ContentResult",
+            // TODO: Other main ones?
```

Comment:
The main ones would be ObjectResult, StatusCodeResult, and JsonResult (legacy). 90% of the other action result types inherit from one of the first two.

Reply:
Good call - added these in!